File: src/dotnet/Harness.Engine/PathResolver.cs
namespace Harness.Engine;

/// <summary>
/// Resolves paths relative to the working directory (the repo root, from where the driver
/// invokes the harness), with a fallback to the binary's directory. Shared by whatever
/// injects files into the prompt (skills, docs).
/// </summary>
public static class PathResolver
{
    public static string Resolve(string path)
    {
        var trimmed = path.Trim();
        if (Path.IsPathRooted(trimmed))
            return trimmed;

        var cwd = Directory.GetCurrentDirectory();
        var fromCwd = Path.GetFullPath(trimmed, cwd);
        if (Exists(fromCwd) && IsContained(fromCwd, cwd))
            return fromCwd;

        var baseDir = AppContext.BaseDirectory;
        var fromBase = Path.GetFullPath(trimmed, baseDir);
        if (Exists(fromBase) && IsContained(fromBase, baseDir))
            return fromBase;

        // Neither the CWD nor the BaseDirectory worked — missing, or a symlink steering the
        // target outside both authorized bases. Returns the original joined path WITHOUT
        // following the link (not the resolved target, which would be outside the base);
        // the caller's subsequent File.Exists naturally fails when the target isn't
        // actually accessible/authorized.
        return fromBase;
    }

    private static bool Exists(string path) => File.Exists(path) || Directory.Exists(path);

    /// <summary>
    /// Symlink containment (RFC §6.3): resolves the link's final target (if <paramref name="candidate"/>
    /// is one) and checks it's actually inside <paramref name="baseDir"/>, comparing
    /// canonical paths by real directory prefix — not lexical string prefix (which would
    /// let "/base-evil" pass as contained in "/base").
    /// </summary>
    private static bool IsContained(string candidate, string baseDir)
    {
        var normalizedBase = Path.GetFullPath(baseDir).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
        var target = Path.GetFullPath(ResolveFinalTarget(candidate))
            .TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);

        var comparison = OperatingSystem.IsWindows() ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;
        if (target.Equals(normalizedBase, comparison))
            return true;

        var baseWithSep = normalizedBase + Path.DirectorySeparatorChar;
        return target.StartsWith(baseWithSep, comparison);
    }

    /// <summary>Follows the link (if any) to its final target; returns the path itself if it's not a link or resolution fails (e.g. a broken link).</summary>
    private static string ResolveFinalTarget(string path)
    {
        try
        {
            if (File.Exists(path))
                return File.ResolveLinkTarget(path, returnFinalTarget: true)?.FullName ?? path;

            if (Directory.Exists(path))
                return Directory.ResolveLinkTarget(path, returnFinalTarget: true)?.FullName ?? path;
        }
        catch
        {
            // Broken/inaccessible link: treated as unresolved — falls back to the original
            // path, and the containment check above decides based on it.
        }

        return path;
    }
}

File: src/go/harnessengine/path_resolver.go
package harnessengine

import (
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

// ResolvePath resolves a relative path against the working directory (repo root, from
// where the driver invokes the harness), falling back to the binary's directory. Shared by
// anything that injects files into the prompt (skills, docs).
func ResolvePath(path string) string {
	trimmed := strings.TrimSpace(path)
	if filepath.IsAbs(trimmed) {
		return trimmed
	}

	if cwd, err := os.Getwd(); err == nil {
		fromCwd := filepath.Join(cwd, trimmed)
		if pathExists(fromCwd) && isContained(fromCwd, cwd) {
			if canonical, err := filepath.EvalSymlinks(fromCwd); err == nil {
				return canonical
			}
			return fromCwd
		}
	}

	baseDir := binaryDir()
	fromBase := filepath.Join(baseDir, trimmed)
	if pathExists(fromBase) && isContained(fromBase, baseDir) {
		if canonical, err := filepath.EvalSymlinks(fromBase); err == nil {
			return canonical
		}
		return fromBase
	}

	// Neither the CWD nor the binary dir served — absent, or a symlink diverting the
	// target outside both authorized bases. Returns the original join path WITHOUT
	// following the link (not the resolved target, which would be outside the base); the
	// caller's subsequent existence check naturally fails when the target isn't reachable.
	return fromBase
}

func binaryDir() string {
	exe, err := os.Executable()
	if err != nil {
		if cwd, err := os.Getwd(); err == nil {
			return cwd
		}
		return "."
	}
	dir := filepath.Dir(exe)
	if canonical, err := filepath.EvalSymlinks(dir); err == nil {
		return canonical
	}
	return dir
}

func pathExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

// isContained checks symlink containment (RFC §6.3): resolves candidate's final target (if
// it is a symlink) and confirms it is actually inside baseDir, comparing canonical paths by
// real directory prefix — not lexical string prefix (which would let "/base-evil" pass as
// contained in "/base").
func isContained(candidate, baseDir string) bool {
	normalizedBase, err := filepath.Abs(baseDir)
	if err != nil {
		normalizedBase = baseDir
	}
	normalizedBase = strings.TrimRight(normalizedBase, string(filepath.Separator))

	target := resolveFinalTarget(candidate)
	if abs, err := filepath.Abs(target); err == nil {
		target = abs
	}
	target = strings.TrimRight(target, string(filepath.Separator))

	if runtime.GOOS == "windows" {
		normalizedBase = strings.ToLower(normalizedBase)
		target = strings.ToLower(target)
	}

	if target == normalizedBase {
		return true
	}

	return strings.HasPrefix(target, normalizedBase+string(filepath.Separator))
}

// resolveFinalTarget follows the link (if any) to its final target; returns the path
// itself if it is not a link or resolution fails (e.g. broken link).
func resolveFinalTarget(path string) string {
	resolved, err := filepath.EvalSymlinks(path)
	if err != nil {
		return path
	}
	return resolved
}
